### PR TITLE
Deprecated Index.get_duplicates()

### LINKED ISSUE #20239: DEPR: Index.get_duplicates
completely duplicates (pun-intended) .duplicated / .drop_duplicates and is unecessary.

```
In [1]: pd.Index([1, 2, 3, 2, 3, 4, 3]).get_duplicates()
Out[1]: [2, 3]

In [2]: pd.Index([1, 2, 3, 2, 3, 4, 3]).drop_duplicates()
Out[2]: Int64Index([1, 2, 3, 4], dtype='int64')

In [3]: pd.Index([1, 2, 3, 2, 3, 4, 3]).duplicated()
Out[3]: array([False, False, False,  True,  True, False,  True])

In [4]: i = pd.Index([1, 2, 3, 2, 3, 4, 3])

In [5]: i[i.duplicated()]
Out[5]: Int64Index([2, 3, 3], dtype='int64')

In [6]: i[i.duplicated()].unique()
Out[6]: Int64Index([2, 3], dtype='int64')
```

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #20239
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -2061,6 +2061,11 @@ def test_cached_properties_not_settable(self):
         with tm.assert_raises_regex(AttributeError, "Can't set attribute"):
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: need to remove the usage from all tests or catch the warnings

```
(pandas) bash-3.2$ grep -r get_duplicates pandas
pandas/core/reshape/concat.py:                overlap = concat_index.get_duplicates()
Binary file pandas/core/reshape/__pycache__/concat.cpython-36.pyc matches
Binary file pandas/core/__pycache__/frame.cpython-36.pyc matches
pandas/core/frame.py:            duplicates = index.get_duplicates()
Binary file pandas/core/indexes/__pycache__/datetimelike.cpython-36.pyc matches
Binary file pandas/core/indexes/__pycache__/base.cpython-36.pyc matches
pandas/core/indexes/datetimelike.py:    def get_duplicates(self):
pandas/core/indexes/datetimelike.py:        values = Index.get_duplicates(self)
pandas/core/indexes/base.py:    def get_duplicates(self):
pandas/core/indexes/base.py:        >>> pd.Index([1, 2, 2, 3, 3, 3, 4]).get_duplicates()
pandas/core/indexes/base.py:        >>> pd.Index([1., 2., 2., 3., 3., 3., 4.]).get_duplicates()
pandas/core/indexes/base.py:        >>> pd.Index(['a', 'b', 'b', 'c', 'c', 'c', 'd']).get_duplicates()
pandas/core/indexes/base.py:        >>> pd.Index(dates).get_duplicates()
pandas/core/indexes/base.py:        >>> pd.Index([1, 2, 3, 2, 3, 4, 3]).get_duplicates()
pandas/core/indexes/base.py:        >>> pd.Index([1, 2, 3, 4]).get_duplicates()
pandas/core/indexes/base.py:        >>> pd.Index(dates).get_duplicates()
pandas/core/indexes/base.py:    _get_duplicates = get_duplicates
Binary file pandas/tests/indexes/__pycache__/test_multi.cpython-36-PYTEST.pyc matches
Binary file pandas/tests/indexes/datetimes/__pycache__/test_datetime.cpython-36-PYTEST.pyc matches
pandas/tests/indexes/datetimes/test_datetime.py:    def test_get_duplicates(self):
pandas/tests/indexes/datetimes/test_datetime.py:        result = idx.get_duplicates()
Binary file pandas/tests/indexes/timedeltas/__pycache__/test_timedelta.cpython-36-PYTEST.pyc matches
pandas/tests/indexes/timedeltas/test_timedelta.py:    def test_get_duplicates(self):
pandas/tests/indexes/timedeltas/test_timedelta.py:        result = idx.get_duplicates()
pandas/tests/indexes/test_multi.py:            assert mi.get_duplicates() == []
pandas/tests/indexes/test_multi.py:                assert mi.get_duplicates() == []
```

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -1870,14 +1873,20 @@ def get_duplicates(self):
...
         DatetimeIndex([], dtype='datetime64[ns]', freq=None)
         """
+        warnings.warn("'get_duplicates' is deprecated and will be removed in "
+                      "a future release. You can use "
+                      "idx[idx.duplicated()].unique() instead",
+                      FutureWarning, stacklevel=2)
+
+        return self._get_duplicates()
+
+    def _get_duplicates(self):
```
Comment (jreback) [inline pandas/core/indexes/base.py]: would rather change the impl itself here as well to the suggested one

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
Hmm I made that change but it looks like this actually doesn't work for MultiIndexes

```python
mi = pd.MultiIndex.from_arrays([['a', 'b'], ['c', 'd']])
mi[mi.duplicated()].unique()
*** TypeError: Cannot infer number of levels from empty list
```

Not being as familiar on the MI side of things, do you think it's worth creating that as a separate issue and merging before this (I think it's a problem with the `unique` method there) or do you have another suggestion on how to handle?

**Comment by jreback** [comment]
this is a bug in ``.unique()`` on an empty MI

```
In [13]: pd.MultiIndex.from_arrays([[], []])
Out[13]: 
MultiIndex(levels=[[], []],
           labels=[[], []])

In [14]: pd.MultiIndex.from_arrays([[], []]).unique()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-14-d5be3b23b8c3> in <module>()
----> 1 pd.MultiIndex.from_arrays([[], []]).unique()

~/pandas/pandas/core/indexes/multi.py in unique(self, level)
   1071 
   1072         if level is None:
-> 1073             return super(MultiIndex, self).unique()
   1074         else:
   1075             level = self._get_level_number(level)

~/pandas/pandas/core/indexes/base.py in unique(self, level)
   4366             self._validate_index_level(level)
   4367         result = super(Index, self).unique()
-> 4368         return self._shallow_copy(result)
   4369 
   4370     def drop_duplicates(self, keep='first'):

~/pandas/pandas/core/indexes/multi.py in _shallow_copy(self, values, **kwargs)
    558             # discards freq
    559             kwargs.pop('freq', None)
--> 560             return MultiIndex.from_tuples(values, **kwargs)
    561         return self.view()
    562 

~/pandas/pandas/core/indexes/multi.py in from_tuples(cls, tuples, sortorder, names)
   1315             if names is None:
   1316                 msg = 'Cannot infer number of levels from empty list'
-> 1317                 raise TypeError(msg)
   1318             arrays = [[]] * len(names)
   1319         elif isinstance(tuples, (np.ndarray, Index)):

TypeError: Cannot infer number of levels from empty list

```

let's fix that one first, i'll create an issue.

**Comment by toobaz** [comment]
> Hmm I made that change but it looks like this actually doesn't work for MultiIndexes

@WillAyd : fixed, can you rebase?

File: pandas/tests/indexes/test_multi.py [hunk 3]
```diff
@@ -2413,7 +2413,12 @@ def check(nlevels, with_nulls):
         for a in [101, 102]:
             mi = MultiIndex.from_arrays([[101, a], [3.5, np.nan]])
             assert not mi.has_duplicates
-            assert mi.get_duplicates() == []
+
+            with warnings.catch_warnings(record=True):
+                # Deprecated - see GH20239
+                assert mi.get_duplicates().equals(MultiIndex.from_arrays(
```
Comment (WillAyd) [inline pandas/tests/indexes/test_multi.py]: Had to tweak the assertion here given the return value is no longer a list, though I assume you are aware of that from the original issue. With that said, this is a different behavior for non-datetimelikes (which were returning a like-Index object) - is it worth documenting that in the whatsnew or is this getting into too technical of a distinction?

Comment (jreback) [reply pandas/tests/indexes/test_multi.py]: this is fine

**Comment by jreback** [comment]
can you rebase

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
can you just check that we don't have any warnings showing from this

**Comment by WillAyd** [comment]
Checked the output of `test_fast.sh` and there were no warnings so I think this is OK - lmk if I should be checking somewhere else

**Comment by jreback** [comment]
thanks @WillAyd nice patch!

**Pull request merged**